Controllers/HomeController.cs

```diff
@@ -3,21 +3,89 @@
 using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Mvc;
+using System.Text.RegularExpressions;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using Newtonsoft.Json;
+using System.Dynamic;
+using Newtonsoft.Json.Linq;
 
 namespace Zoltu.Bags.Client.Controllers
 {
 	public class HomeController : Controller
 	{
+		private static readonly Regex productRegex = new Regex(@"product\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex tagRegex = new Regex(@"tags\/(.*?)\/*?$", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex minPriceRegex = new Regex(@"minprice\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex maxPriceRegex = new Regex(@"maxprice\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex aboutRegex = new Regex(@"aboutus", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+		// Best practices: https://www.asp.net/web-api/overview/advanced/calling-a-web-api-from-a-net-client
+		private static readonly HttpClient client = new HttpClient();
+
+		static HomeController()
+		{
+			client.BaseAddress = new Uri("https://bags-api.zoltu.com/api/");
+			client.DefaultRequestHeaders.Accept.Clear();
+			client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+		}
+
 		[Route("")]
 		[Route("app/{*path}")]
-		public IActionResult App(string path = null)
+		public async Task<IActionResult> App(String path = null)
 		{
-			ViewData["url"] = "https://bagcupid.com/";
-			ViewData["type"] = "website";
-			ViewData["title"] = "Bag Cupid";
-			ViewData["description"] = "What is your dream bag? Are you having trouble finding it? Let us help you!";
+			var url = "https://bagcupid.com/";
+			var type = "website";
+			var title = "Bag Cupid";
+			var description = "What is your dream bag? Are you having trouble finding it? Let us help you!";
+			var image = "https://bagcupid.com/img/logo/bagcupid.png";
+
+			if (!String.IsNullOrEmpty(path))
+			{
+				var productId = productRegex.Match(path).Groups[1].Value?.TryParseUInt64();
+				var tagId = tagRegex.Match(path).Groups[1].Value;
+				var minPrice = minPriceRegex.Match(path).Groups[1].Value.TryParseUInt64();
+				var maxPrice = maxPriceRegex.Match(path).Groups[1].Value.TryParseUInt64();
+				var aboutUs = aboutRegex.Match(path).Success;
+
+				if (productId.HasValue)
+				{
+					var product = await CallApi($"products/{productId}");
+
+					url = $"https://bagcupid.com/{path}";
+
+					if (product?.images != null && product.images.Count > 0)
+						image = (string)(((JObject)product)["images"].Aggregate((minItem, nextItem) => (int)minItem["priority"] < (int)nextItem["priority"] ? minItem : nextItem))["large"];
```

Comment:
To break it, we probably need implementation of IComparable, which might be even more complex. So my recommendation is to be this function with comment to describe its purpose.

Casting is already eliminated in your defined type "Product", as conversion is held. When large will not be coming in JSON, it will have null value.